add GVEC as MHD optimizable

Rykath opened the pull request:
This PR adds an `Optimizable`-class for the *Galerkin Variational Equilibrium Code* ([GVEC](https://gitlab.mpcdf.mpg.de/gvec-group/gvec)). GVEC is a flexible 3D ideal MHD equilibrium solver inspired by VMEC with a few notable differences. The two most prominent are:
* The radial direction is discretized using B-Splines rather than finite differences on a fixed grid. This allows representing the region of the magnetic axis accurately and evaluations anywhere in the domain.
* The coordinate mapping from flux aligned coordinates to real space is flexible and not fixed to $(R,Z,\phi)$. This also allows using [a flexible curve-following coordinate frame](https://doi.org/10.1088/1361-6587/adba11), e.g. to compute equilibria for a [figure-8 stellarator](https://iopscience.iop.org/article/10.1088/1361-6587/adb64b)

GVEC is developed at the division of Numerical Methods of the Max Planck Institute for Plasma Physics in Garching, Germany. It has recently been released publicly and open source (MIT License).
Details on the implementation of GVEC are partially described in the [documentation](https://gvec.readthedocs.io).

---

This PR adds a `simsopt.mhd.gvec.Gvec` optimizable, similar to the `Vmec` and `Spec` optimizables.

* DoFs: `phiedge`, `boundary`, `pressure` and `iota` or `current` profile
* GVEC is parallelized with OpenMP and MPI. For now, parallel `least_squares_mpi_solve` is only supported with one MPI process for each GVEC run.
* fixed boundary
* prescribing a `iota` or (toroidal) `current` profile
* basic return functions for `iota`, `iota_axis`, `volume`, etc.
* evaluating other equilibrium quantities (e.g. boozer transform) can use GVEC's python API, the `GVECQuantity` optimizable can be used to access any of the computable quantities of GVEC.

This PR is still work in progress, but we wanted to show our progress as soon as possible. Open ToDo's:
* [x] prescribe toroidal current
* [ ] documentation
* [x] examples / stellarator benchmarks
  * [x] `1DOF_circularCrossSection_varyAxis_targetIota_gvec.py`
  * [x] `1DOF_circularCrossSection_varyR0_targetVolume_gvec.py`
  * [x] `2DOF_gvecOnly_targetIotaAndVolume.py`
* [ ] unit tests
* [x] add `SurfaceGVECFourier` to represent surface DoFs when not using cylindrical coordinates
  * this is not a `Surface` object, e.g. does not support the `gamma()` method

Let us know what else you need.

Rykath committed:
add GVEC optimizable

adds a GVEC optimizable class, similar to the VMEC optimizable

Rykath committed:
enable gvec run_stages

* refactor Gvec optimizable to use the run_stages script: allows running GVEC with refinement and current constraint (optimization using picard iterations)
* profile objects are now available as iota_profile, pressure_profiel and current_profile, while iota and I_tor are return functions
* Gvec init using dictionary of parameters, from_parameter_file classmethod additionally
* adapted example parameter file & run script for 1DoF circularCrossSection target Volume

Rykath committed:
init GVEC optimizable from boundary and profile objects

* from_parameters & from_parameter_file classmethods are also available
* allow setting iota & current profile afterwards
* add default parameters (minimization and radial resolution)
* the varyR0_targetVolume example now no longer needs an input file

smiet reviewed (changes requested):
Looking very nice! Only major issue is with the use of SurfaceRZFourier as the boundary object, which I believe should be replaced with a `SurfaceGVEC` or `SurfaceRZFourierOnFrame` lest users try to access, plot, evaluate, and optimize properties of the boundary directly. 


Aditionally, unit tests need to be written to test all the new code. See tests/mhd/vmec.py for a good example to adapt. 
This will also involve updating the workflow files to install GVEC.

smiet commented on examples/stellarator_benchmarks/1DOF_circularCrossSection_varyR0_targetVolume_gvec.py:
```diff
@@ -0,0 +1,87 @@
+
+assert np.abs(surf.get_rc(0, 0) - 0.7599088773175) < 1.0e-5
+assert np.abs(equil.volume() - 0.15) < 1.0e-6
+assert np.abs(surf.volume() - 0.15) < 1.0e-6
```
You are evaluating the surfaces' volume, but this is only equal to equil.volume in this case, and is misleading. See other comments on creating a SurfaceGVEC class for representing the surface.

Rykath committed:
fix typing annotations for python < 3.10

Rykath committed:
fix run() with force and style changes

Rykath committed:
overhaul GVEC optimizable

* improved handling of parameters:
  * surface can be specified via the parameters or as an object
  * profiles can be specified via the parameters, an object or a constant
* iota profile is always used, but if current is specified it is only used as the initial guess
* removed from_parameters constructor - all features now available in init
* add logger property to access the GVEC logger
* make paths in the parameterfile absolute when writing parameters for each run

Rykath committed:
update with 'master'

Rykath committed:
small fixes

Rykath committed:
update gvec v1.1

Rykath committed:
overhaul Gvec optimizable

* new argument `keep_gvec_intermediates`: passed to gvec.run with values "all", "stages" or None
* new decorator `return_function` to mark all return functions which are then registered with `return_fn_map`
* cleaned return functions
  * vacuum_well is now the vacuum magnetic well depth with the same  formual as vmec
  * removed current & iota return functions that returned arrays as the resolution is not customizable
  * return functions now use the .state().evaluate(...)
* new function `boundary_to_params` & overhaul of `boundary_from_params`: now supports non-stellarator-symmetry
* fix current optimization by setting `picard_current` correctly

Rykath committed:
fix boundary_to_params

Rykath committed:
fix stellsym in boundary_from_params

Rykath committed:
update with 'master'

Rykath committed:
add more examples

Rykath committed:
revise gvec return functions, state & prototype restart

* Gvec.state is now a property and not a return function
* new Gvec.beta_avg return function
* new MirrorRatio optimizable based on gvec
* new Elongation optimizable based on gvec
* new Gvec.from_rundir constructor
* Gvec.run_count now starts at 0
* changed Gvec option restart_file -> restart
* prototype for restart functionality (wip)

Rykath committed:
add keep_failures option & fixes

Rykath committed:
fix keep_failures

Rykath committed:
rework return functions

Rykath committed:
fix 2DOF_gvecOnly & adjust gvec version

Rykath commented:
I apologize for not updating this in a very long while.
In the meantime we've done quite some work behind the scenes, in particular extending the utilities and improving the stability of GVEC, which was limiting this integration. With GVEC `v1.3` I think we have everything ready on the GVEC side, and only the open ToDos here remain.

Rykath committed:
add SurfaceGVECFourier

Rykath committed:
add from/to_gvec_parameters methods

for SurfaceRZFourier & SurfaceGVECFourier

Rykath committed:
update with 'master'

Rykath committed:
move GVECSurfaceDoFs & add autoscale method

Rykath committed:
test & fix to/from_gvec_parameters

* new gvec parameter file for testing: LandremanPaul2021_QA (converted from vmec input)
* added import guard clause for qsc for test_surface_rzfourier

Rykath committed:
add tests for gvec initialization

Rykath committed:
add GVEC to CI

Rykath committed:
add GVECQuantity.rms & sfl argument

* allows quickly evaluating the root-mean-square
* allows targeting straight-fieldline quantities

Rykath committed:
cleanup recompute_flag

Rykath committed:
fix rundir & automatic rundir deletion

Rykath committed:
fix formatting

Rykath committed:
add gvec tests current & pressure profile

Rykath committed:
fix SurfaceRZFourier.to_gvec_parameters

Rykath committed:
remove surfacegvec

was previously moved to mhd.gvec_dofs but accidently and should have been deleted

Rykath committed:
add QA lowres testcase

Rykath committed:
disabled test_run_from_rundir

Rykath committed:
update with 'master'

Rykath committed:
fix rundir deletion

Rykath committed:
fix axis initialization

Rykath committed:
fix tests
